README.md
[![npm version](https://badge.fury.io/js/%40spacecomx%2Fcdk-organization-billing-alarm.svg)](https://badge.fury.io/js/%40spacecomx%2Fcdk-organization-billing-alarm)
[![PyPI version](https://badge.fury.io/py/spacecomx.cdk-organization-billing-alarm.svg)](https://badge.fury.io/py/spacecomx.cdk-organization-billing-alarm)

# @spacecomx/cdk-organization-billing-alarm

Multi-account CDK construct to monitor estimated billing charges with alerts and notifications for a AWS Organization. It gives you the capability to monitor specific AWS Service charges, by a linked AWS account in a master/payer account. It can create customizable billing alarms for multiple linked AWS accounts within AWS Organization. It can subscribe multiple email address endpoints to an SNS Topic created by the package or it can use an existing SNS Topic Arn within the master/payer account.

For single AWS account billing alarm requirements, see [@spacecomx/cdk-billing-alarm](https://github.com/spacecomx/cdk-billing-alarm#readme)

## Table of Contents

- [Prerequisites](#prerequisites)
- [Installation](#installation)
- [Usage](#usage)

  - [Example 1: Create billing alarm in a master/payer account](#example_1)
  - [Example 2: Link alarm to an existing SNS Topic Arn](#example_2)
  - [Example 3: Link alarm to specific AWS Account in master/payer account](#example_3)
  - [Example 4: Updating/Removing SNS Topic email address endpoint subscription](#example_4)

- [Post Deployment](#post-deployment)
- [API Documentation](#api-documentation)

## Prerequisites

> :warning: Before you can create a billing alarm, you must enable billing alerts in your master/payer account when using consolidated billing. For more information, see [Enabling Billing Alerts](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/monitor_estimated_charges_with_cloudwatch.html#turning_on_billing_metrics).

## Installation

TypeScript/JavaScript:

```bash
npm i @spacecomx/cdk-organization-billing-alarm
```

or:

```bash
yarn add @spacecomx/cdk-organization-billing-alarm
```

Python:

```bash
pip install spacecomx.cdk-organization-billing-alarm
```

## Usage

<a name="example_1"></a>

### Example 1: Create billing alarm in a master/payer account.

This type of billing alarm configuration will provide estimated charges for every AWS Service that you use, in addition to the estimated overall total of your AWS charges within your AWS master/payer account e.g (AWS Organization).

- It creates a new SNS Topic in your AWS master/payer account.
- It then subscribes the email address endpoint to the topic.
- It then creates the Cloudwach estimate billing alarm.
- It then associates the SNS topic with the newly created billing alarm.

> :small_orange_diamond: The `emailAddress` is an endpoint that subscribes to a SNS topic. The `thresholdAmount` is the amount in USD, that will trigger the alarm when AWS charges exceed the threshold.

```typescript
import { Stack, StackProps } from '@aws-cdk/core';
import {
  MasterAccountAlarm,
  MasterAccountAlarmProps,
} from '@spacecomx/cdk-organization-billing-alarm';

export class OrganizationBillingAlarmStack extends Stack {
  constructor(scope: Construct, id: string, props: StackProps) {
    super(scope, id, props);

    const config: MasterAccountAlarmProps = {
      alarmConfiguration: {
        topicDescription: 'Organization Billing Alarm Topic',
        emailAddress: ['billing@example.org', 'admin@example.org'],
        alarmDescription: 'Consolidated Billing Alarm: All AWS Services',
        thresholdAmount: 140,
      },
    };

    new MasterAccountAlarm(this, 'MasterAccountAlarm', config);
  }
}
```

<a name="example_2"></a>

### Example 2: Link alarm to an existing SNS Topic Arn.

This type of billing alarm configuration simliar to [example 1](#example_1), with the exception that the billing alarm is now linked to an existing SNS Topic in the master/payer account.

- It uses an existing SNS Topic within the AWS master/payer account.
- It imports a secret by secret name that was created in master/payer account using AWS Secrets Manager e.g (prod/billing/topicArn).
- It then subscribes the email address endpoint to the retrieved SNS Topic's Arn.
- It then creates the Cloudwach estimate billing alarm.
- It then associates the SNS topic with the newly created billing alarm.

> :small_orange_diamond: The `secretName` option is required and used to retrieve the existing SNS Topic's Arn from AWS Secrets Manager. The secret name **must exist** in the same account and region as the master/payer AWS account.

```typescript
import { Stack, StackProps } from '@aws-cdk/core';
import {
  MasterAccountAlarm,
  MasterAccountAlarmProps,
} from '@spacecomx/cdk-organization-billing-alarm';

export class OrganizationBillingAlarmStack extends Stack {
  constructor(scope: Construct, id: string, props: StackProps) {
    super(scope, id, props);

    const config: MasterAccountAlarmProps = {
      secretName: 'prod/billing/topicArn', // existing sns topic arn
      alarmConfiguration: {
        emailAddress: ['john@example.org'],
        alarmDescription: 'Consolidated Billing Alarm: All AWS Services',
        thresholdAmount: 140,
      },
    };

    new MasterAccountAlarm(this, 'MasterAccountAlarm', config);
  }
}
```

<a name="example_3"></a>

### Example 3: Link alarm to specific AWS Account in master/payer account.

This type of billing alarm configuration provides customizable options for linked AWS accounts within the master/payer account. The packages `accountConfiguration` required option allows for multiple AWS accounts to link to a single SNS Topic using the required `secretName` option.

- It uses an existing SNS Topic's Arn within the AWS master/payer account.
- It imports a secret by secret name that was created in master/payer account using AWS Secrets Manager e.g (prod/billing/topicArn).
- It then creates the Cloudwach estimate billing alarm associated to the linked AWS account.
- It then associates the SNS topic with the newly created billing alarm.

> :small_orange_diamond: The `secretName` option is required and used to retrieve the existing SNS Topic's Arn from AWS Secrets Manager. The secret name **must exist** in the same account and region as the master/payer AWS account.

```typescript
import { Stack, StackProps } from '@aws-cdk/core';
import {
  LinkedAccountAlarm,
  LinkedAccountAlarmProps,
} from '@spacecomx/cdk-organization-billing-alarm';

export class OrganizationBillingAlarmStack extends Stack {
  constructor(scope: Construct, id: string, props: StackProps) {
    super(scope, id, props);

    const config: LinkedAccountAlarmProps = {
      secretName: 'prod/billing/topicArn',
      accountConfiguration: [
        {
          account: '444455556666',
          alarmName: 'Billing Alarm (Acc: 444455556666)', // named by aws-cdk (recommended)
          alarmDescription:
            'Billing Alarm: All AWS Services (Acc: 444455556666)',
          thresholdAmount: 50,
        },
      ],
    };

    new LinkedAccountAlarm(this, 'LinkedAccountAlarm', config);
  }
}
```

-OR-

You can manage **multiple linked AWS accounts** each with its own billing alarm configuration, with a single implementation of the `LinkedAccountAlarm` construct. The code example below uses a **_single SNS Topic_** to manage alerts and notifications for all linked accounts within the master/payer account. Each linked account can send alerts and notifications to either a single email address or multiple email addresses, subscribed to the SNS Topic within the master/payer account

> :small_orange_diamond: Should you need more that one SNS topic e.g. you want seperate out linked accounts by department or business unit, each having its own SNS Topic. One option would be to, firstly create an SNS Topic in the master/payer account for each department or business unit. Remember to create a new secret with that SNS Topic's Arn in AWS Secrets Manager. Then simply, new up `LinkedAccountAlarm` constructs that group those linked accounts by department or business unit, each with its own SNS Topic.

```typescript
const config: LinkedAccountAlarmProps = {
  secretName: 'prod/billing/other/topicArn', // single topic used by multiple accounts
  accountConfiguration: [
    {
      account: '444455556666',
      alarmDescription: 'Billing Alarm: All AWS Services (Acc: 444455556666)',
      thresholdAmount: 50,
    },
    {
      account: '123456789000',
      alarmDescription: 'Billing Alarm: All AWS Services (Acc: 123456789000)',
      thresholdAmount: 120,
    },
  ],
};
```

> :small_orange_diamond: You can also link the alarm to a specific AWS Service, per linked AWS account. Use the `awsService` option. See the code example below.

```typescript
const config: LinkedAccountAlarmProps = {
  secretName: 'prod/billing/topicArn',
  accountConfiguration: [
    {
      account: '444455556666',
      alarmDescription: 'Billing Alarm: All AWS Services (Acc: 444455556666)',
      thresholdAmount: 50,
    },
    {
      account: '123456789000',
      alarmDescription: 'Billing Alarm: Amazon DynamoDB (Acc: 123456789000)',
      thresholdAmount: 120,
      awsService: 'AmazonDynamoDB', // alert on Amazon DynamoDB charges
    },
  ],
};
```

> :small_orange_diamond: Should you need to remove a `awsService` subscribed to a specific linked account, simply remove the `awsService` option.

<a name="example_4"></a>

### Example 4: Updating/Removing SNS Topic email address endpoint subscription.

When would you use this option?

- When an existing SNS topic that already has an email address endpoint subscribed to it,
- or you dont require any additional email address endpoints to be added to the existing topic,
- or you want to add your own email address endpoints manually to the SNS topic (not recommended),
- or you want to remove an email address endpoints from the SNS topic created with the package,
- or you want to add or update email address endpoints for the SNS topic using the package.

> :small_orange_diamond: To manage adding and removing endpoints for the SNS topic, use the packages `emailAddress` option. You can simply set the `emailAddress: ['john@example.org']` option to `emailAddress: []`. Please note that email endpoint subscriptions created manually via AWS SNS Console **will not be removed** by the package.

```typescript
const config: MasterAccountAlarmProps = {
  alarmConfiguration: {
    emailAddress: [], // required
    ...
  },
};
```

Adding and removing endpoints means **_you will need to again confirm the subscription_** of each email address you specified with `emailAddress` option or those added manually by you e.g AWS SNS console (not recommended). See [post deployment](#post-deployment) for details.

> :warning: Please be **cautious**. Without an endpoint been provided i.e. (email address been subscribed to the SNS topic), the billing alarm will still trigger when exceeding the alarm threshold. However you will **not recieve any email alarm notifications** via email.

## Post Deployment

Once the Billing Alarm Stack resources has been successfully created in your AWS account, you will need to confirm the subscription of each email address you specified with the `emailAddress` configuration option. Clicking on the **"Confirm Subscription"** link for that email, will automatically activate billing alarm notifications for that email address.

If you did not receive the email, you can process a **"Request Confirmation"** for the subscription from the Simple Notification Service (SNS) console within your AWS account.

> :warning: Without confirming the email subscription you will **not recieve any email alarm notifications** via email. The billing alarm will trigger when exceeding the alarm threshold, but **you will not be notified** via email.

## API Documentation

For more detail, see [API documentation](https://github.com/spacecomx/cdk-organization-billing-alarm/blob/main/API.md)

## Contributions

Contributions of all kinds are welcome! Check out our [contributor's guide](https://github.com/spacecomx/cdk-organization-billing-alarm/blob/main/CONTRIBUTING.md) and our [code of conduct](https://github.com/spacecomx/cdk-organization-billing-alarm/blob/main/CODE_OF_CONDUCT.md)

## Credits

- [Wayne Gibson](https://github.com/waynegibson)

## Alternatives

- [@spacecomx/cdk-billing-alarm](https://github.com/spacecomx/cdk-billing-alarm#readme) - used for single account billing alarm requirements and some.

## License

@spacecomx/cdk-organization-billing-alarm is distributed under the [MIT](https://github.com/spacecomx/cdk-organization-billing-alarm/blob/main/LICENSE) license.


CHANGELOG.md
# Changelog

## [v1.0.6](https://github.com/spacecomx/cdk-organization-billing-alarm/tree/v1.0.6) (2021-08-12)

[Full Changelog](https://github.com/spacecomx/cdk-organization-billing-alarm/compare/v1.0.5...v1.0.6)

**Merged pull requests:**

- chore\(deps\): upgrade dependencies [\#5](https://github.com/spacecomx/cdk-organization-billing-alarm/pull/5) ([github-actions[bot]](https://github.com/apps/github-actions))
- chore\(deps\): upgrade dependencies [\#4](https://github.com/spacecomx/cdk-organization-billing-alarm/pull/4) ([github-actions[bot]](https://github.com/apps/github-actions))
- chore\(deps\): upgrade dependencies [\#3](https://github.com/spacecomx/cdk-organization-billing-alarm/pull/3) ([github-actions[bot]](https://github.com/apps/github-actions))

## [v1.0.5](https://github.com/spacecomx/cdk-organization-billing-alarm/tree/v1.0.5) (2021-08-04)

[Full Changelog](https://github.com/spacecomx/cdk-organization-billing-alarm/compare/v1.0.4...v1.0.5)

**Merged pull requests:**

- chore\(deps\): upgrade dependencies [\#2](https://github.com/spacecomx/cdk-organization-billing-alarm/pull/2) ([github-actions[bot]](https://github.com/apps/github-actions))

## [v1.0.4](https://github.com/spacecomx/cdk-organization-billing-alarm/tree/v1.0.4) (2021-08-03)

[Full Changelog](https://github.com/spacecomx/cdk-organization-billing-alarm/compare/v1.0.3...v1.0.4)

**Merged pull requests:**

- chore\(deps\): upgrade dependencies [\#1](https://github.com/spacecomx/cdk-organization-billing-alarm/pull/1) ([github-actions[bot]](https://github.com/apps/github-actions))

## [v1.0.3](https://github.com/spacecomx/cdk-organization-billing-alarm/tree/v1.0.3) (2021-08-02)

[Full Changelog](https://github.com/spacecomx/cdk-organization-billing-alarm/compare/v1.0.2...v1.0.3)

## [v1.0.2](https://github.com/spacecomx/cdk-organization-billing-alarm/tree/v1.0.2) (2021-08-02)

[Full Changelog](https://github.com/spacecomx/cdk-organization-billing-alarm/compare/v1.0.1...v1.0.2)

## [v1.0.1](https://github.com/spacecomx/cdk-organization-billing-alarm/tree/v1.0.1) (2021-08-02)

[Full Changelog](https://github.com/spacecomx/cdk-organization-billing-alarm/compare/v1.0.0...v1.0.1)

## [v1.0.0](https://github.com/spacecomx/cdk-organization-billing-alarm/tree/v1.0.0) (2021-08-02)

[Full Changelog](https://github.com/spacecomx/cdk-organization-billing-alarm/compare/v1.0.0-beta.0...v1.0.0)

## [v1.0.0-beta.0](https://github.com/spacecomx/cdk-organization-billing-alarm/tree/v1.0.0-beta.0) (2021-08-02)

[Full Changelog](https://github.com/spacecomx/cdk-organization-billing-alarm/compare/05d1c1202831313dc46f5f8e8799267172f7bcb0...v1.0.0-beta.0)



\* *This Changelog was automatically generated by [github_changelog_generator](https://github.com/github-changelog-generator/github-changelog-generator)*


test/snapshot.test.ts
import '@aws-cdk/assert/jest';
import { SynthUtils } from '@aws-cdk/assert';
import {
  MasterAccountAlarmTest,
  MasterAccountAlarmWithAwsServiceTest,
  LinkedAccountAlarmTest,
} from '../src/integ.default';

test('integration snapshot test for master account billing alarm configuation', ()=> {
  const integration = new MasterAccountAlarmTest();

  integration.stack.forEach(stack => {
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  });
});

test('integration snapshot test for master account billing alarm configuation with aws service', ()=> {
  const integration = new MasterAccountAlarmWithAwsServiceTest();

  integration.stack.forEach(stack => {
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  });
});

test('integration snapshot test for linked account billing alarm configuation', ()=> {
  const integration = new LinkedAccountAlarmTest();

  integration.stack.forEach(stack => {
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  });
});

test/linked-account-alarm.test.ts
import { countResources, expect as expectCDK, haveResourceLike } from '@aws-cdk/assert';
import * as cdk from '@aws-cdk/core';
import { LinkedAccountAlarm, LinkedAccountAlarmProps } from '../src';

test('ensure resources exist to create a linked account alarm with existing sns topic', () => {
  const app = new cdk.App();
  const stack = new cdk.Stack(app, 'TestStack');

  const config: LinkedAccountAlarmProps = {
    secretName: 'test/billing/topicArn',
    accountConfiguration: [
      { account: '444455556666', alarmName: 'Billing Alarm (All Services)', alarmDescription: 'Consolidated billing alarm for all AWS service charge estimates (Account: 444455556666)', thresholdAmount: 50 },
    ],
  };

  new LinkedAccountAlarm(stack, 'LinkedBillingAlarm', config);

  expectCDK(stack).to(countResources('AWS::SNS::Topic', 0));
  expectCDK(stack).to(countResources('AWS::SNS::Subscription', 0));

  expectCDK(stack).to(countResources('AWS::CloudWatch::Alarm', 1));
  expectCDK(stack).to(
    haveResourceLike('AWS::CloudWatch::Alarm', {
      AlarmActions: [
        {
          'Fn::Join': [
            '',
            [
              '{{resolve:secretsmanager:arn:',
              {
                Ref: 'AWS::Partition',
              },
              ':secretsmanager:',
              {
                Ref: 'AWS::Region',
              },
              ':',
              {
                Ref: 'AWS::AccountId',
              },
              ':secret:test/billing/topicArn:SecretString:::}}',
            ],
          ],
        },
      ],
      AlarmName: 'Billing Alarm (All Services)',
      AlarmDescription: 'Consolidated billing alarm for all AWS service charge estimates (Account: 444455556666)',
      ComparisonOperator: 'GreaterThanOrEqualToThreshold',
      EvaluationPeriods: 1,
      MetricName: 'EstimatedCharges',
      Namespace: 'AWS/Billing',
      Period: 21600,
      Statistic: 'Maximum',
      Threshold: 50,
      Dimensions: [
        {
          Name: 'Currency',
          Value: 'USD',
        },
        {
          Name: 'LinkedAccount',
          Value: '444455556666',
        },
      ],
    }),
  );
});

test('ensure resources exist to create a linked account alarm with existing sns topic with email subscription', () => {
  const app = new cdk.App();
  const stack = new cdk.Stack(app, 'TestStack');

  const config: LinkedAccountAlarmProps = {
    secretName: 'test/billing/topicArn',
    accountConfiguration: [
      { account: '444455556666', alarmDescription: 'Consolidated billing alarm for all AWS service charge estimates (Account: 444455556666)', thresholdAmount: 50, emailAddress: ['john@example.org'] },
    ],
  };

  new LinkedAccountAlarm(stack, 'LinkedBillingAlarm', config);

  expectCDK(stack).to(countResources('AWS::SNS::Topic', 0));

  expectCDK(stack).to(countResources('AWS::SNS::Subscription', 1));
  expectCDK(stack).to(
    haveResourceLike('AWS::SNS::Subscription', {
      Endpoint: 'john@example.org',
      Protocol: 'email',
      TopicArn: {
        'Fn::Join': [
          '',
          [
            '{{resolve:secretsmanager:arn:',
            {
              Ref: 'AWS::Partition',
            },
            ':secretsmanager:',
            {
              Ref: 'AWS::Region',
            },
            ':',
            {
              Ref: 'AWS::AccountId',
            },
            ':secret:test/billing/topicArn:SecretString:::}}',
          ],
        ],
      },
    }),
  );

  expectCDK(stack).to(countResources('AWS::CloudWatch::Alarm', 1));
  expectCDK(stack).to(
    haveResourceLike('AWS::CloudWatch::Alarm', {
      AlarmDescription: 'Consolidated billing alarm for all AWS service charge estimates (Account: 444455556666)',
      ComparisonOperator: 'GreaterThanOrEqualToThreshold',
      EvaluationPeriods: 1,
      MetricName: 'EstimatedCharges',
      Namespace: 'AWS/Billing',
      Period: 21600,
      Statistic: 'Maximum',
      Threshold: 50,
      Dimensions: [
        {
          Name: 'Currency',
          Value: 'USD',
        },
        {
          Name: 'LinkedAccount',
          Value: '444455556666',
        },
      ],
    }),
  );
});

test('ensure resources exist to create a linked account alarm with existing sns topic with specific aws service', () => {
  const app = new cdk.App();
  const stack = new cdk.Stack(app, 'TestStack');

  const config: LinkedAccountAlarmProps = {
    secretName: 'test/billing/topicArn',
    accountConfiguration: [
      { account: '444455556666', alarmDescription: 'Billing Alarm for AWS DynamoDB charge estimates only (Account: 444455556666)', thresholdAmount: 50, awsService: 'AmazonDynamoDB' },
    ],
  };

  new LinkedAccountAlarm(stack, 'LinkedBillingAlarm', config);

  expectCDK(stack).to(countResources('AWS::SNS::Topic', 0));
  expectCDK(stack).to(countResources('AWS::SNS::Subscription', 0));

  expectCDK(stack).to(countResources('AWS::CloudWatch::Alarm', 1));
  expectCDK(stack).to(
    haveResourceLike('AWS::CloudWatch::Alarm', {
      AlarmDescription: 'Billing Alarm for AWS DynamoDB charge estimates only (Account: 444455556666)',
      ComparisonOperator: 'GreaterThanOrEqualToThreshold',
      EvaluationPeriods: 1,
      MetricName: 'EstimatedCharges',
      Namespace: 'AWS/Billing',
      Period: 21600,
      Statistic: 'Maximum',
      Threshold: 50,
      Dimensions: [
        {
          Name: 'Currency',
          Value: 'USD',
        },
        {
          Name: 'LinkedAccount',
          Value: '444455556666',
        },
        {
          Name: 'ServiceName',
          Value: 'AmazonDynamoDB',
        },
      ],
    }),
  );
});

test('ensure resources exist to create a billing alarm for multiple linked accounts', () => {
  const app = new cdk.App();
  const stack = new cdk.Stack(app, 'TestStack');

  const config: LinkedAccountAlarmProps = {
    secretName: 'test/billing/topicArn',
    accountConfiguration: [
      { account: '444455556666', alarmDescription: 'Consolidated billing alarm for all AWS service charge estimates (Account: 444455556666)', thresholdAmount: 50 },
      { account: '123456789000', alarmDescription: 'Consolidated billing alarm for all AWS service charge estimates (Account: 123456789000)', thresholdAmount: 120 },
    ],
  };

  new LinkedAccountAlarm(stack, 'LinkedBillingAlarm', config);

  expectCDK(stack).to(countResources('AWS::SNS::Topic', 0));
  expectCDK(stack).to(countResources('AWS::SNS::Subscription', 0));

  expectCDK(stack).to(countResources('AWS::CloudWatch::Alarm', 2));
  expectCDK(stack).to(
    haveResourceLike('AWS::CloudWatch::Alarm', {
      AlarmDescription: 'Consolidated billing alarm for all AWS service charge estimates (Account: 444455556666)',
      ComparisonOperator: 'GreaterThanOrEqualToThreshold',
      EvaluationPeriods: 1,
      MetricName: 'EstimatedCharges',
      Namespace: 'AWS/Billing',
      Period: 21600,
      Statistic: 'Maximum',
      Threshold: 50,
      Dimensions: [
        {
          Name: 'Currency',
          Value: 'USD',
        },
        {
          Name: 'LinkedAccount',
          Value: '444455556666',
        },
      ],
    }),
  );
  expectCDK(stack).to(
    haveResourceLike('AWS::CloudWatch::Alarm', {
      AlarmDescription: 'Consolidated billing alarm for all AWS service charge estimates (Account: 123456789000)',
      ComparisonOperator: 'GreaterThanOrEqualToThreshold',
      EvaluationPeriods: 1,
      MetricName: 'EstimatedCharges',
      Namespace: 'AWS/Billing',
      Period: 21600,
      Statistic: 'Maximum',
      Threshold: 120,
      Dimensions: [
        {
          Name: 'Currency',
          Value: 'USD',
        },
        {
          Name: 'LinkedAccount',
          Value: '123456789000',
        },
      ],
    }),
  );
});

test/master-account-alarm.test.ts
import { countResources, expect as expectCDK, haveResourceLike } from '@aws-cdk/assert';
import * as cdk from '@aws-cdk/core';
import { MasterAccountAlarm, MasterAccountAlarmProps } from '../src';

test('ensure resources exist to create a master account alarm with new sns topic', () => {
  const app = new cdk.App();
  const stack = new cdk.Stack(app, 'TestStack');

  const config: MasterAccountAlarmProps = {
    alarmConfiguration: {
      topicDescription: 'Organizational billing alarm topic',
      emailAddress: ['john@example.org'],
      alarmDescription: 'Consolidated billing alarm for all AWS Service charges',
      thresholdAmount: 140,
    },
  };

  new MasterAccountAlarm(stack, 'MasterBillingAlarm', config);

  expectCDK(stack).to(countResources('AWS::SNS::Topic', 1));
  expectCDK(stack).to(
    haveResourceLike('AWS::SNS::Topic', {
      DisplayName: 'Organizational billing alarm topic',
    }),
  );

  expectCDK(stack).to(countResources('AWS::SNS::Subscription', 1));
  expectCDK(stack).to(
    haveResourceLike('AWS::SNS::Subscription', {
      TopicArn: { Ref: 'MasterBillingAlarmTopic11ACFFE5' },
      Endpoint: 'john@example.org',
      Protocol: 'email',
    }),
  );

  expectCDK(stack).to(countResources('AWS::CloudWatch::Alarm', 1));
  expectCDK(stack).to(
    haveResourceLike('AWS::CloudWatch::Alarm', {
      AlarmActions: [
        { Ref: 'MasterBillingAlarmTopic11ACFFE5' },
      ],
      AlarmDescription: 'Consolidated billing alarm for all AWS Service charges',
      ComparisonOperator: 'GreaterThanOrEqualToThreshold',
      EvaluationPeriods: 1,
      MetricName: 'EstimatedCharges',
      Namespace: 'AWS/Billing',
      Period: 21600,
      Statistic: 'Maximum',
      Threshold: 140,
      Dimensions: [
        {
          Name: 'Currency',
          Value: 'USD',
        },
      ],
    }),
  );
});

test('ensure resources exist to create a master account alarm with specific aws service', () => {
  const app = new cdk.App();
  const stack = new cdk.Stack(app, 'TestStack');

  const config: MasterAccountAlarmProps = {
    alarmConfiguration: {
      topicDescription: 'Organizational billing alarm topic',
      emailAddress: ['john@example.org'],
      alarmDescription: 'Billing Alarm for AWS DynamoDB charge estimates only (Account: 12345444000)',
      thresholdAmount: 140,
      awsService: 'AmazonDynamoDB',
    },
  };

  new MasterAccountAlarm(stack, 'MasterBillingAlarm', config);

  expectCDK(stack).to(countResources('AWS::SNS::Topic', 1));
  expectCDK(stack).to(countResources('AWS::SNS::Subscription', 1));

  expectCDK(stack).to(countResources('AWS::CloudWatch::Alarm', 1));
  expectCDK(stack).to(
    haveResourceLike('AWS::CloudWatch::Alarm', {
      AlarmActions: [
        { Ref: 'MasterBillingAlarmTopic11ACFFE5' },
      ],
      AlarmDescription: 'Billing Alarm for AWS DynamoDB charge estimates only (Account: 12345444000)',
      ComparisonOperator: 'GreaterThanOrEqualToThreshold',
      EvaluationPeriods: 1,
      MetricName: 'EstimatedCharges',
      Namespace: 'AWS/Billing',
      Period: 21600,
      Statistic: 'Maximum',
      Threshold: 140,
      Dimensions: [
        {
          Name: 'Currency',
          Value: 'USD',
        },
        {
          Name: 'ServiceName',
          Value: 'AmazonDynamoDB',
        },
      ],
    }),
  );
});

test('ensure resources exist to create a master account alarm using existing sns topic', () => {
  const app = new cdk.App();
  const stack = new cdk.Stack(app, 'TestStack');

  const config: MasterAccountAlarmProps = {
    secretName: 'test/billing/topicArn',
    alarmConfiguration: {
      emailAddress: ['john@example.org'],
      alarmDescription: 'Consolidated organisation AWS Services',
      thresholdAmount: 140,
    },
  };

  new MasterAccountAlarm(stack, 'MasterBillingAlarm', config);

  expectCDK(stack).to(countResources('AWS::SNS::Topic', 0));

  expectCDK(stack).to(countResources('AWS::SNS::Subscription', 1));
  expectCDK(stack).to(
    haveResourceLike('AWS::SNS::Subscription', {
      Endpoint: 'john@example.org',
      Protocol: 'email',
    }),
  );

  expectCDK(stack).to(countResources('AWS::CloudWatch::Alarm', 1));
  expectCDK(stack).to(
    haveResourceLike('AWS::CloudWatch::Alarm', {
      AlarmActions: [
        {
          'Fn::Join': ['', ['{{resolve:secretsmanager:arn:',
            {
              Ref: 'AWS::Partition',
            },
            ':secretsmanager:',
            {
              Ref: 'AWS::Region',
            },
            ':',
            {
              Ref: 'AWS::AccountId',
            },
            ':secret:test/billing/topicArn:SecretString:::}}']],
        },
      ],
    }),
  );
});


test('ensure resources exist to create a master account alarm using existing sns topic without email subscription', () => {
  const app = new cdk.App();
  const stack = new cdk.Stack(app, 'TestStack');

  const config: MasterAccountAlarmProps = {
    secretName: 'test/billing/topicArn',
    alarmConfiguration: {
      emailAddress: [],
      alarmDescription: 'Consolidated billing alarm for all AWS Service charges',
      thresholdAmount: 140,
    },
  };

  new MasterAccountAlarm(stack, 'MasterBillingAlarm', config);

  expectCDK(stack).to(countResources('AWS::SNS::Topic', 0));
  expectCDK(stack).to(countResources('AWS::SNS::Subscription', 0));

  expectCDK(stack).to(countResources('AWS::CloudWatch::Alarm', 1));
  expectCDK(stack).to(
    haveResourceLike('AWS::CloudWatch::Alarm', {
      AlarmActions: [
        {
          'Fn::Join': ['', ['{{resolve:secretsmanager:arn:',
            {
              Ref: 'AWS::Partition',
            },
            ':secretsmanager:',
            {
              Ref: 'AWS::Region',
            },
            ':',
            {
              Ref: 'AWS::AccountId',
            },
            ':secret:test/billing/topicArn:SecretString:::}}']],
        },
      ],
      AlarmDescription: 'Consolidated billing alarm for all AWS Service charges',
      ComparisonOperator: 'GreaterThanOrEqualToThreshold',
      EvaluationPeriods: 1,
      MetricName: 'EstimatedCharges',
      Namespace: 'AWS/Billing',
      Period: 21600,
      Statistic: 'Maximum',
      Threshold: 140,
      Dimensions: [
        {
          Name: 'Currency',
          Value: 'USD',
        },
      ],
    }),
  );
});

.projenrc.js
const { AwsCdkConstructLibrary, ProjectType, NpmAccess } = require('projen');

const RELEASE_STATUS = true;
const RELEASE_BRANCH = 'main';
const PRE_RELEASE = '';
const MAYOR_VERSION = 1;
const RELEASE_TO_NPM = true;
const AWS_CDK_VERSION = '1.118.0';

const project = new AwsCdkConstructLibrary({
  name: 'cdk-organization-billing-alarm',
  packageName: '@spacecomx/cdk-organization-billing-alarm',
  description:
    'Multi-account CDK construct to monitor estimated billing charges with alerts and notifications for a AWS Organization. It gives you the capability to monitor specific AWS Service charges, by a linked AWS account in a master/payer account',
  author: 'Wayne Gibson',
  authorAddress: 'wayne.gibson@spacecomx.com',
  repositoryUrl:
    'https://github.com/spacecomx/cdk-organization-billing-alarm.git',
  keywords: [
    'aws',
    'cdk',
    'aws-constructs',
    'constructs',
    'aws-billing',
    'billing',
    'billing-alarm',
    'aws-organizations',
    'cloudwatch',
    'aws-cloudwatch',
    'cloudwatch-alarms',
    'sns',
    'spacecomx',
  ],
  license: 'MIT',
  copyrightOwner: 'Spacecomx LLC',
  projectType: ProjectType.LIB,

  cdkVersion: AWS_CDK_VERSION,
  cdkDependencies: [
    '@aws-cdk/core',
    '@aws-cdk/aws-sns',
    '@aws-cdk/aws-secretsmanager',
  ],

  deps: ['@spacecomx/cdk-billing-alarm'],
  peerDeps: ['@spacecomx/cdk-billing-alarm'],

  release: RELEASE_STATUS,
  defaultReleaseBranch: RELEASE_BRANCH,
  prerelease: PRE_RELEASE,
  majorVersion: MAYOR_VERSION,

  releaseToNpm: RELEASE_TO_NPM,
  npmAccess: NpmAccess.PUBLIC,

  publishToPypi: {
    distName: 'spacecomx.cdk-organization-billing-alarm',
    module: 'spacecomx.cdk_organization_billing-alarm',
  },
});

const exclude = [
  'cdk.out',
  'cdk.context.json',
  'yarn-error.log',
  'dependabot.yml',
  '.env',
];

project.gitignore.exclude(...exclude);
project.npmignore.exclude(...exclude);

project.synth();


src/integ.default.ts
import * as cdk from '@aws-cdk/core';
import { MasterAccountAlarm, MasterAccountAlarmProps, LinkedAccountAlarm, LinkedAccountAlarmProps } from './index';

export class MasterAccountAlarmTest {
  readonly stack: cdk.Stack[];

  constructor() {
    const app = new cdk.App();

    const env = {
      region: process.env.CDK_DEFAULT_REGION,
      account: process.env.CDK_DEFAULT_ACCOUNT,
    };

    const stack = new cdk.Stack(app, 'TestStack', { env });

    const config: MasterAccountAlarmProps = {
      alarmConfiguration: {
        topicDescription: 'Organizational billing alarm topic',
        emailAddress: ['john@example.org'],
        alarmDescription: 'Consolidated billing alarm for all AWS Service charges',
        thresholdAmount: 140,
        awsService: 'AmazonDynamoDB',
      },
    };

    new MasterAccountAlarm(stack, 'MasterAccountAlarm', config);

    this.stack = [stack];
  }
}
export class MasterAccountAlarmWithAwsServiceTest {
  readonly stack: cdk.Stack[];

  constructor() {
    const app = new cdk.App();

    const env = {
      region: process.env.CDK_DEFAULT_REGION,
      account: process.env.CDK_DEFAULT_ACCOUNT,
    };

    const stack = new cdk.Stack(app, 'TestStack', { env });

    const config: MasterAccountAlarmProps = {
      alarmConfiguration: {
        topicDescription: 'Organizational billing alarm topic',
        emailAddress: ['john@example.org'],
        alarmDescription: 'Billing Alarm for AWS DynamoDB charge estimates only (Account: 12345444000)',
        thresholdAmount: 140,
        awsService: 'AmazonDynamoDB',
      },
    };

    new MasterAccountAlarm(stack, 'MasterAccountAlarm', config);

    this.stack = [stack];
  }
}

export class LinkedAccountAlarmTest {
  readonly stack: cdk.Stack[];

  constructor() {
    const app = new cdk.App();

    const env = {
      region: process.env.CDK_DEFAULT_REGION,
      account: process.env.CDK_DEFAULT_ACCOUNT,
    };

    const stack = new cdk.Stack(app, 'TestStack', { env });

    const config: LinkedAccountAlarmProps = {
      secretName: 'test/billing/topicArn',
      accountConfiguration: [
        { account: '444455556666', alarmDescription: 'Consolidated billing alarm for all AWS service charge estimates (Account: 444455556666)', thresholdAmount: 50, emailAddress: ['john@example.org'] },
        { account: '123456789000', alarmDescription: 'Billing Alarm for AWS DynamoDB charge estimates only (Account: 123456789000)', thresholdAmount: 120, awsService: 'AmazonDynamoDB' },
      ],
    };

    new LinkedAccountAlarm(stack, 'LinkedAccountAlarm', config);

    this.stack = [stack];
  }
}
